### DISCUSSION TITLE
Do we have the optimal cache settings for most users by default?

### DISCUSSION THREAD

**Discussion by Pentaphon** [transmission/transmission#5183] [General]
Has this community ever taken a good hard look at the default cache settings, done any testing, and come up with optimal values that will work for most people with a modern PC? With 4.x, is it time to revisit this and come up with an optimal cache setting?

The current setting is: 

`cache-size-mb: 4`

 qBittorrent's libtorrent backend, for example, seems to use the OS cache by default. Should we do the same for modern systems?

**Comment by ilikepeaches** [comment]
How big might the OS cache (that libtorrent is using) tend to be?  i.e. what's the scale of the difference here?

Has anything about this setting or functionality changed from Transmission 3 to 4?  i.e. could it be the cause of the transfer performance decrease discussed in https://github.com/transmission/transmission/issues/5003?

**Reply by Pentaphon** [reply]
> How big might the OS cache (that libtorrent is using) tend to be? i.e. what's the scale of the difference here?
> 
> Has anything about this setting or functionality changed from Transmission 3 to 4? i.e. could it be the cause of the transfer performance decrease discussed in #5003?

Depends on the OS, depends on the RAM the system has, etc. No way to tell. All we know is that libtorrent uses OS cache by default so that might be a good direction to take for the future of Transmission, unless somebody can think of something better.

**Comment by ckerr** [comment]
> qBittorrent's libtorrent backend, for example, seems to use the OS cache by default. Should we do the same for modern systems?

Maybe?

The goal was to group together contiguous blocks together after downloading them so that, instead of writing them to disk one-by-one, they could be written together in a single write. For example the "SSD write limit" problem is not as bad as it was 5 years ago but still, writing a big media to disk in 16KB chunks is a lot of writes.

It's true that longer writes will cause the libtransmission thread to block until disk writes are done async (there is already a very long ticket open for this feature) but IMO the write cache is still worthwhile.

**Reply by Pentaphon** [reply]
> It's true that longer writes will cause the libtransmission thread to block until disk writes are done async (there is already a very long ticket open for this feature) but IMO the write cache is still worthwhile.

Is using the OS cache the most optimal solution or do you think there's a better way to do it these days?

**Reply by ckerr** [reply]
What would you suggest @Pentaphon?

**Reply by Pentaphon** [reply]
I think the OS cache is the best way forward but I was wondering if somebody more knowledgeable than myself could offer a better solution. Should I just make an issue to implement using the OS cache in case somebody out there could contribute a PR?

**Reply by ckerr** [reply]
When you say "implement using the OS cache" wdym specifically?

**Reply by Pentaphon** [reply]
>enable_os_cache
    Files are opened normally, with the OS caching reads and writes.
disable_os_cache
    This opens all files in no-cache mode. This corresponds to the OS not letting blocks for the files linger in the cache. This makes sense in order to avoid the bittorrent client to potentially evict all other processes' cache by simply handling high throughput and large files. If libtorrent's read cache is disabled, enabling this may reduce performance.
https://www.libtorrent.org/reference-Settings.html

This basically but I don't have the knowledge to explain libtorrent's documentation much further than linking it to you.

**Reply by andreygursky** [reply]
Even in default blocking mode ```write()``` puts data not onto the disk but to the OS cache. As long as you write slower than the OS can write the actual data to the disk (so that "dirty" data in cache wouldn't exceed the allowed maximum), I wouldn't expect any difference in blocking. But. ```write()``` is a system call. So if you choose to write 4 MB piece by calling ```write()``` 256 times each 16 KB or for the whole 4 MB piece only once, the latter should definitively have less overhead (costly context switches) and thus preferred. Since write cache has not only size constraints but also time constrains, in case of rather slower downloads small chunks in OS cache will be eventually forced to be written to the disk, resulting in needless disk operations. Thus again writing a piece at once seems to be better. And
> The goal was to group together contiguous blocks together after downloading them ....

seems to me to be still a pretty good approach.

BTW, may I ask here about the read cache behavior of transmission (2.92+)? If a peer asks blocks in a 4 MB piece (and the speed is not very high to serve it in, say, just <1s), would transmission read the whole piece at once, so that the further disk access is avoided (both for single blocks and background read-ahead of the data after this piece) during sending those blocks?

**Reply by ckerr** [reply]
> BTW, may I ask here about the read cache behavior of transmission (2.92+)? If a peer asks blocks in a 4 MB piece (and the speed is not very high to serve it in, say, just <1s), would transmission read the whole piece at once, so that the further disk access is avoided (both for single blocks and background read-ahead of the data after this piece) during sending those blocks?

Good question. Reading is always done in block-sized chunks, not piece-sized chunks.

If (by coincidence) someone requests a block that we just got from somewhere else & it's still in the write cache, we pull it from there.

Otherwise, we tell the OS that we're going to need that block soon via `tr_sys_file_fadvise(..., TR_SYS_FILE_ADVICE_WILL_NEED)` which is implemented in [file-posix.cc](https://github.com/transmission/transmission/blob/main/libtransmission/file-posix.cc) and [file-win32.cc](https://github.com/transmission/transmission/blob/main/libtransmission/file-win32.cc).

**Reply by andreygursky** [reply]
Just a notice: now it's gone (#6332).

**Comment by dantearmok** [comment]
Balancing dynamic loads is never easy.

fwiw, I like your choice as-is.  Let the OS do the full buffering.  Async will be a good improvement tho.

- Dan.

**Comment by Pentaphon** [comment]
@tearfur I noticed that you posted here https://github.com/transmission/transmission/issues/5261#issuecomment-1599071744 about having good results downloading with a new commit and included your settings such as a cache value of 16 rather than our default of 4. Thoughts on whether 16  or some other value should be the default cache value?

**Reply by tearfur** [reply]
I have no idea, but I can try with cache value 4 and report the results.

**Reply by tearfur** [reply]
https://github.com/transmission/transmission/issues/5261#issuecomment-1599965426
